### PR TITLE
Code review for api-example

### PR DESCRIPTION
**Pull request by jwhui**

### CODE REVIEW

File: examples/apps/cli/main.c [hunk 1]
```diff
@@ -108,12 +152,43 @@ int main(int argc, char *argv[])
...
+    otSetStateChangedCallback(sInstance, handleNetifStateChanged, sInstance);
+
+    /* Register GPIO button handlers for first 2 buttons */
+    otSysGpioRegisterCallback(BUTTON_GPIO_PORT, BUTTON_1_PIN, handleGpioInterrupt1, sInstance);
+    otSysGpioIntEnable(BUTTON_GPIO_PORT, BUTTON_1_PIN);
+
+    otSysGpioRegisterCallback(BUTTON_GPIO_PORT, BUTTON_2_PIN, handleGpioInterrupt2, sInstance);
+    otSysGpioIntEnable(BUTTON_GPIO_PORT, BUTTON_2_PIN);
+
+    /* Disable ICMP Echo handling so we can process it from the ip6 handler.
```
Comment (jwhui) [inline examples/apps/cli/main.c]: I think this example will be a lot simpler/cleaner if we just keep UDP and remove ICMPv6 Echo Request. UDP is also the preferred path for applications to use. Thoughts?

Comment (Vyrastas) [reply examples/apps/cli/main.c]: If that is preferred, I have no problem just sticking with UDP.  That will shorten the Codelab a little bit too which is always a good thing.

Comment (jwhui) [reply examples/apps/cli/main.c]: Having looked at the code, I think the UDP-only path is overall simpler and cleaner. I realize now there's a number of hoops necessary to support ICMPv6. Also, users on the forums generally want to see examples of UDP and/or CoAP, not really how to intercept ICMPv6 packets.

Comment (Vyrastas) [reply examples/apps/cli/main.c]: SGTM, let's do that then

File: examples/platforms/nrf52840/gpio.c [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+void in_pin1_handler(uint32_t pin, nrf_gpiote_polarity_t action)
+{
+    (void)pin;
+    (void)action;
+
+    /* Clear button interrupt */
+    otSysGpioIntClear(NRF_GPIOTE_EVENTS_PORT, pin);
+
+    /* Call the Button 1 handler registered by the application */
+    mHandler1(sInstance);
```
Comment (jwhui) [inline examples/platforms/nrf52840/gpio.c:72]: This is being called from interrupt context. OpenThread APIs are designed to be called from a single thread of execution. In the bare-metal example, the preferred approach is to set a flag, and later have something like `otSysGpioProcess()` called from the main loop, which then checks the flag and calls the callback at that time.

**Pull request closed without merging**